libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1027 @@
+  return g_object_new (FLATPAK_TYPE_REMOTE_REF,
+                       "remote-name", collection_id,
+                       "name", app_name,
+                       "kind", kind,
+                       NULL);
+}
+
+static FlatpakRemoteRef *
+flatpak_remote_ref_from_uninstall_action_detail (JsonObject  *detail,
+                                                 GError     **error)
+{
+  const gchar *app_name = NULL;
+  FlatpakRefKind kind;
+
+  if (!parse_flatpak_ref_from_detail (detail, &app_name, &kind, error))
+    return NULL;
+
+  return g_object_new (FLATPAK_TYPE_REMOTE_REF,
+                       "remote-name", "none",
+                       "name", app_name,
+                       "kind", kind,
+                       NULL);
+}
+
+static FlatpakRemoteRef *
+flatpak_remote_ref_from_action_detail (EosUpdaterUtilFlatpakRemoteRefActionType   action_type,
+                                       JsonObject                                *detail,
+                                       GError                                   **error)
+{
+  switch (action_type)
+    {
+      case EUU_FLATPAK_REMOTE_REF_ACTION_INSTALL:
+        return flatpak_remote_ref_from_install_action_detail (detail, error);
+      case EUU_FLATPAK_REMOTE_REF_ACTION_UNINSTALL:
+        return flatpak_remote_ref_from_uninstall_action_detail (detail, error);
+      default:
+        g_assert_not_reached ();
+    }
+}
+
+static FlatpakRemoteRefAction *
+flatpak_remote_ref_action_from_json_node (JsonNode *node,
+                                          GError **error)
+{
+  const gchar *action_type_str = NULL;
+  JsonObject *object = json_node_get_object (node);
+  JsonObject *detail_object = NULL;
+  g_autoptr(FlatpakRemoteRef) flatpak_remote_ref = NULL;
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(JsonNode) serial_node = NULL;
+  gint64 serial;
+  EosUpdaterUtilFlatpakRemoteRefActionType action_type;
+
+  if (object == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected entry node to be an object, but was %s", node_str);
+
+      return NULL;
+    }
+
+  action_type_str = json_object_get_string_member (object, "action");
+
+  if (action_type_str == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected an 'action' member in the autoinstall spec (at %s)", node_str);
+
+      return NULL;
+    }
+
+  if (!flatpak_remote_ref_action_type_parse (action_type_str, &action_type, error))
+    return NULL;
+
+  /* XXX: How do we validate that this value is sensible, as far as I can
+   * tell, json_object_get_int_member would return 0 if the member was not
+   * set or was not an int */
```

Comment:
I think the code you have is OK here, feel free to remove the XXX (unless you file an issue against json-glib for better/missing API for this purpose)